Add support for rectangular area lights (RectAreaLight) #47

mrxz opened the pull request:
Initial version of supporting (rectangular) area lights. I tried to keep the changes straightforward and clear. I've also added an example. Apologies for it being a bit lacklustre, but it does demonstrate how much quicker the image converges due to NEE and MIS. Though, the benefit of MIS on top of NEE does appear to be limited in this scene. 

Not sure about the naming of some of the methods and variables, so any comments or suggestions are appreciated. The handling of the environment map can be further improved. It's now half-way in between being a light type (uniformly picked along with the area lights for NEE), yet at the same time still distinct. 

Couple of remarks (in no particular order):
- The `THREE.RectAreaLight` emits light from the back face. I've implemented it this way in the shader as well.
- The existing `misHeuristic` contained a bug and didn't actually implement the power heuristic, but accidentally the balance heuristic (`a^2 / (a^2 + b^2)` was accidentially `a^2 / (a^2 + ab)` which simplifies to just `a / (a + b)` :slightly_smiling_face: 
- Next-event estimation picks uniformly between any of the lights or the env map.

**Area lights (with MIS and NEE 293 samples)**
![image](https://user-images.githubusercontent.com/8823461/173152178-35f797c4-2db5-4623-a20c-95da67cd3fc3.png)
**Area lights (without MIS and NEE 1400 samples)**
![image](https://user-images.githubusercontent.com/8823461/173152143-f508be43-85e5-4d55-bcb4-2789d248a8cf.png)
**Area lights (without MIS and NEE 50k samples)**
![image](https://user-images.githubusercontent.com/8823461/173152693-2bfe3faa-fe11-417b-b72f-b0995d38fe3a.png)

gkjohnson commented:
Thanks @mrxz! This looks great. I still have to take a deep dive into the code but I'm having trouble running the example to start. When pulling this branch and even just running the "materialBall" example my chrome tab / webgl crashes and I see this in the log:

```
THREE.WebGLProgram: Program Info Log: C:\fakepath(844,1-6): warning X4000: use of potentially uninitialized variable (dyn_index_vec3)
internal error: unexpected input register type
internal error: failed to emit instruction

Warning: D3D shader compilation failed with default flags. (ps_5_0)
 Retrying with skip validation
C:\fakepath(844,1-6): warning X4000: use of potentially uninitialized variable (dyn_index_vec3)
internal error: unexpected input register type
internal error: failed to emit instruction

Warning: D3D shader compilation failed with skip validation flags. (ps_5_0)
 Retrying with skip optimization
C:\fakepath(844,1-6): warning X4000: use of potentially uninitialized variable (dyn_index_vec3)
C:\fakepath(1319,1): warning X4000: use of potentially uninitialized variable (f_intersectsRectangle)
C:\fakepath(1320,1): warning X4000: use of potentially uninitialized variable (f_intersectsRectangle)
C:\fakepath(1321,1): warning X4000: use of potentially uninitialized variable (f_intersectsRectangle)
C:\fakepath(1673,1): warning X4000: use of potentially uninitialized variable (f_envMapDirectionPdf)
C:\fakepath(1510,1): warning X4000: use of potentially uninitialized variable (f_transmissionPDF__SurfaceRec)
C:\fakepath(1400,1): warning X4000: use of potentially uninitialized variable (f_ggxDistribution)
C:\fakepath(1673,1): warning X4000: use of potentially uninitialized variable (f_envMapDirectionPdf)
C:\fakepath(844,1-6): warning X4000: use of potentially uninitialized variable (dyn_index_vec3)
C:\fakepath(1400,1): warning X4000: use of potentially uninitialized variable (f_ggxDistribution)
C:\fakepath(1510,1): warning X4000: use of potentially uninitialized variable (f_transmissionPDF__SurfaceRec)
C:\fakepath(1400,1): warning X4000: use of potentially uninitialized variable (f_ggxDistribution)
C:\fakepath(844,1-6): warning X4000: use of potentially uninitialized variable (dyn_index_vec3)
C:\fakepath(1400,1): warning X4000: use of potentially uninitialized variable (f_ggxDistribution)
C:\fakepath(1510,1): warning X4000: use of potentially uninitialized variable (f_transmissionPDF__SurfaceRec)
C:\fakepath(1400,1): warning X4000: use of potentially uninitialized variable (f_ggxDistribution)
```

Any thoughts on what might be going on?

mrxz commented:
> Any thoughts on what might be going on?

Interesting, it works for me on both Firefox and Chromium under Linux. Seeing as the error message refers to D3D leads me to think that the GLSL code seems to result in invalid D3D shader code when translated by ANGLE. I'll see if I can reproduce it on a Windows machine and try to solve/workaround it. Curious if this might actually be a bug in ANGLE(?)

mrxz commented:
@gkjohnson Good news, I was able to reproduce it on Windows and have pushed a fixed. The strange thing is that I didn't really change any logic, but just swapped around two branches and inverted the condition... When performing next-event estimation the shader branches to either sample the environment map or one of the (area) lights. It used to sample the environment in the positive clause and the lights in the else clause, but now it's the other way around which magically fixed the issue.

While I can't rule out that there might be something wrong with the code, the fact that it works under Linux and the above change fixes it, feels like a bug in ANGLE somehow. I'll try and dive a bit deeper into to this, when I have time.

gkjohnson commented:
Thanks! It works for me now. I've run into a number of compiler-related errors that don't seem to have an immediate rhyme or reason to them so it's not super surprising that the fix feels a bit nonsensical. It probably is some kind of issue triggered by an optimization or compiler issue.

One thing I notice while testing is that the scenes seem to converge to different results with and without MIS enabled. W/o MIS is definitely the simpler case so there may be some issue with the MIS-case that needs some digging?

| with MIS | without MIS |
|---|---|
| ![image](https://user-images.githubusercontent.com/734200/173212548-7e98614b-a993-4de5-93ba-c8c59c446f19.png) | ![image](https://user-images.githubusercontent.com/734200/173212547-8a798a44-64e0-4a90-a1d1-7c71cb614be2.png) |

mrxz commented:
> W/o MIS is definitely the simpler case so there may be some issue with the MIS-case that needs some digging?

It seems that there is indeed a bug with MIS, which seems present on `main` as well with the environment MIS, albeit less pronounced. I think the pdf is incorrect for specular interactions, as with a perfect mirror you'd expect the pdf to effectively always be 0, unless the NEE ray happens to match a perfect reflection.

See these two screenshots running the code from `main` (so without any area lights), with the floor set as a perfect mirror:
|with MIS|without MIS|
|--|--|
|![image](https://user-images.githubusercontent.com/8823461/173224555-604f2519-c584-414e-8d4c-44a552a79eb3.png)|![image](https://user-images.githubusercontent.com/8823461/173224680-ef816071-3fc5-417c-b54d-64d5e83a2b0b.png)| 

I'll try and figure out what is going wrong. Any ideas or pointers what it might be are welcome 🙂

richardassar commented:
The halos are indeed gone. One remaining issue, which I looked into at depth last night, is a slight residual blur in the MIS case and some fireflies that persist in the non-MIS case. Strangely enough I noticed the opposite behavior (MIS vs non-MIS flipped) in the `materialBall.html` demo versus your `areaLight.html` demo.

The reflection sharpness comes back when removing `envColor` from `gl_FragColor.rgb += attenuatedColor * environmentIntensity * envColor* throughputColor * sampleColor * misWeight / envPdf;`, separately I noticed a discrepancy between the contributions received from `envMapSample` and `sampleEquirectEnvMapColor`, I also traced some issues back to `throughputColor`, adjusting the lines on which it is computed or used fixed this stray samples issue, which sometimes can be seen around the rect area light in the air, eventually these samples are marginalised out. It seems there are some issues at high sample counts where black pixels creep in, drop the `resolutionScale` and increase `samplesPerFrame` in order to see it sooner.

I'll dig into it further today.

richardassar commented:
> Thanks! It works for me now. I've run into a number of compiler-related errors that don't seem to have an immediate rhyme or reason to them so it's not super surprising that the fix feels a bit nonsensical. It probably is some kind of issue triggered by an optimization or compiler issue.
> 
> One thing I notice while testing is that the scenes seem to converge to different results with and without MIS enabled. W/o MIS is definitely the simpler case so there may be some issue with the MIS-case that needs some digging?
> with MIS 	without MIS
> ![image](https://user-images.githubusercontent.com/734200/173212548-7e98614b-a993-4de5-93ba-c8c59c446f19.png) 	![image](https://user-images.githubusercontent.com/734200/173212547-8a798a44-64e0-4a90-a1d1-7c71cb614be2.png)

I had noticed and managed to resolve the discrepancy shown here, will make sure to test for this and include the fix in my next PR.

mrxz commented:
I fixed a small issue where the `envPdf` and `lightPdf` weren't probably scaled based on the amount of lights. Along with the changes from #176 the results look pretty good, but there is still a (clear) visual difference with MIS compared to without MIS in the specular interactions.

I am able to reproduce the issue on `main` with an environment map that approximates an area light (that is very large and very far away). So it's definitely a pre-existing issue, but is very pronounced with area lights at low angles. See the following screenshots of the issue on `main` with the changes from #176, but no area light code (probably want to view them at full size):
| with MIS | without MIS |
| -------| ----|
| ![image](https://user-images.githubusercontent.com/8823461/173246033-7057f07b-dc51-4093-a26f-a7307c1da0f6.png) | ![image](https://user-images.githubusercontent.com/8823461/173246040-a6202c84-660f-4040-b121-35bf3fe6089a.png) |

With MIS the bounced light seems to extend further, whereas without MIS there is a clear higher intensity along with the specular highlight area being "broader" towards the light. Here's a gif that alternates between the two to show the difference:
![specular](https://user-images.githubusercontent.com/8823461/173246692-3dfc38b8-adac-4e7a-9203-71b9be56aa2e.gif)

There also seems to be some 'banding' with MIS enabled in the above example. That might be due to the relatively low resolution environment map (1024x512).

gkjohnson commented:
Maybe we'll see some improvements once #176 is addressed

richardassar commented:
I just pushed a commit that fixes those smaller halos but still see this discrepancy in the area lights demo. 

I had a close look at the MIS vs non-MIS code paths and observed some other differences in output, they occur prior to the perfect reflections fix. I'll continue looking later on.

It should be possible to use some targeted test scenes to isolate these problems but it might also be nice to write some code which tests and visualises various functions and their behaviour under different sampling strategies outside of a full path-trace.

gkjohnson commented:
@mrxz Like you've mentioned with #176 merged things look a lot more consistent, now, though there are still some very noticeable discrepancies. I'll take a closer look at the code this week to see if there's anything I can pull out that might be causing this. Overall it looks great, though - I'm really excited to have this added to the project!

| with MIS | without MIS |
|---|---|
| ![image](https://user-images.githubusercontent.com/734200/173281056-99403990-b4e9-4e47-baf5-a8c0691db279.png) | ![image](https://user-images.githubusercontent.com/734200/173281071-f1610308-321d-490a-bcd5-1daba569ec36.png) |

gkjohnson commented on src/materials/PhysicalPathTracingMaterial.js:
```diff
@@ -378,7 +387,38 @@ export class PhysicalPathTracingMaterial extends MaterialBase {
-						if ( ! bvhIntersectFirstHit( bvh, rayOrigin, rayDirection, faceIndices, faceNormal, barycoord, side, dist ) ) {
+						bool hit = bvhIntersectFirstHit( bvh, rayOrigin, rayDirection, faceIndices, faceNormal, barycoord, side, dist );
+
+						LightSampleRec lightHit = lightsClosestHit( lights, lightCount, rayOrigin, rayDirection );
+
+						if ( lightHit.hit && ( lightHit.dist < dist || !hit ) ) {
```
Do we need to include this light distance check in the "attenuateHit" function, as well?

mrxz replied:
Good catch, I've added it to `attenuateHit` as well.

gkjohnson commented on src/shader/shaderLightSampling.js:
```diff
@@ -0,0 +1,87 @@
+
+LightSampleRec lightsClosestHit( sampler2D lights, uint lightCount, vec3 rayOrigin, vec3 rayDirection ) {
+
+	LightSampleRec lightSampleRec;
+	lightSampleRec.hit = false;
```
Should we initialize all the struct values in case we don't hit any light front faces?

gkjohnson commented on example/areaLight.js:
```diff
@@ -0,0 +1,268 @@
+
+	areaLights = [ areaLight1, areaLight2 ];
+
+	transformControls = new TransformControls( camera, renderer.domElement );
```
Can we add the ability to rotate and scale the light, as well? With the ability to change transform modes using W, E, and R for translate, rotate, and scale respectively?

richardassar replied:
```js
window.addEventListener( 'keydown', function ( event ) {

  switch ( event.keyCode ) {
  
	  case 87: // W
		  transformControls.setMode( 'translate' );
		  break;
  
	  case 69: // E
		  transformControls.setMode( 'rotate' );
		  break;
  
	  case 82: // R
		  transformControls.setMode( 'scale' );
		  break;
  
  }

} ); 
```

gkjohnson reviewed (commented):
Unfortunately nothing stuck out to me when looking at the code regarding the inconsistencies between MIS / Non MIS rendering. Perhaps this is something we can revisit another time?

I think the demo does a good job demonstrating the area light capabilities and we can update it over time! A couple things to adjust - lets get rid of / update the options that don't do anything (emissiveIntensity - maybe it can adjust the primary area light intensity?)

And lets update the docs with the new class information and uniforms.

richardassar commented on src/shader/shaderUtils.js:
```diff
@@ -234,4 +234,42 @@ export const shaderUtils = /* glsl */`
 
 	}
+
+	bool intersectsRectangle( vec3 position, vec3 normal, vec3 u, vec3 v, vec3 rayOrigin, vec3 rayDirection, out float dist ) {
+
+		float dt = dot( rayDirection, normal );
+		float t = ( dot( normal, position ) - dot( normal, rayOrigin ) ) / dt;
+
+		if ( t > EPSILON ) {
+
+			vec3 hit = rayOrigin + rayDirection * t;
+			vec3 vi = hit - position;
+			float a1 = dot( u, vi );
+			if ( a1 >= - 0.5 && a1 <= 0.5 ) {
+
+				float a2 = dot( v, vi );
+				if ( a2 >= - 0.5 && a2 <= 0.5 ) {
+
+					dist = t;
+					return true;
+
+				}
+
+			}
+
+		}
+
+		return false;
+
+	}
```
The computation of `t` can be rewritten as follows, saving a `dot` product.

```glsl
float t = dot(position - rayOrigin, normal) / dot(rayDirection, normal);
```

I've pushed a commit to the spot lights PR that includes this change, it could be factored out into a separate function if it ends up repeated more.

mrxz replied:
I don't really have a reference, but it basically computes the intersection of the ray with the plane as defined by the rect area light (through the `u` and `v` vectors) and then checks if that point falls within the width and height from the center position of the light. I've added a comment describing the process and included a reference for the first part ([ray-plane intersection](https://lousodrome.net/blog/light/2020/07/03/intersection-of-a-ray-and-a-plane/)).

@richardassar Thanks for the optimization. The reference I've added for the ray-plane intersection also mentions this optimization.

gkjohnson commented on example/areaLight.js:
```diff
@@ -0,0 +1,268 @@
+async function init() {
+
+	renderer = new THREE.WebGLRenderer( { antialias: true } );
+	renderer.toneMapping = THREE.ACESFilmicToneMapping;
```
Related to #185 - I'd forgotten to include the line to set the output color space to sRGB in the other examples, so we should add the following here, as well:

`renderer.outputEncoding = THREE.sRGBEncoding`

mrxz commented:
Good news @gkjohnson. I have tracked down the cause of the inconsistencies in the specular interactions. The formula for the `specularPdf` was incorrect (see [e4e2757](https://github.com/gkjohnson/three-gpu-pathtracer/pull/174/commits/e4e275746877aea2579bc6a29792a0137da5bb70)). I did contemplate making a separate PR for the fix, since it strictly speaking isn't related to area lights, but since it's easy to demonstrate with the area lights I added it here.

With this fix, black dots started appearing with MIS was enabled. This was caused by NaN values in the `sampleColor` so I added a check for that as well. Here are the results:
| with MIS | without MIS |
| -- | -- |
| ![image](https://user-images.githubusercontent.com/8823461/173918307-49ada236-58e2-4086-a578-6f7775b08fcc.png) | ![image](https://user-images.githubusercontent.com/8823461/173918334-d6ff0e8f-053e-431d-b79b-b74373f7cb44.png) |

gkjohnson reviewed (commented):
Nice work! A last couple things and we can get this merged!

mrxz commented on src/uniforms/LightsTexture.js:
```diff
@@ -0,0 +1,83 @@
+
+			// u vector
+			l.getWorldQuaternion( worldQuaternion );
+			u.set( l.width, 0, 0 ).applyQuaternion( worldQuaternion );
```
Actually, you'll find that three.js' handling of AreaLights is to _not_ apply the world scale. I was quite surprised by it at first as well. The PR that fixed a bug in the `RectAreaLightHelper` regarding scaling explains the rational behind it: https://github.com/mrdoob/three.js/pull/21848

They adhere to the same behaviour as described in the glTF spec, where properties of lights are unaffected by their world scale. So scaling a node with point lights doesn't 'scale' the point lights, and likewise, scaling a node with area lights doesn't 'scale' the area lights.

gkjohnson replied:
Oh interesting! I didn't realize that had changed - thanks for the info!

mrxz commented on src/materials/PhysicalPathTracingMaterial.js:
```diff
@@ -660,7 +660,7 @@ export class PhysicalPathTracingMaterial extends MaterialBase {
 								vec3 sampleColor;
 								float lightMaterialPdf = bsdfResult( outgoing, normalize( invBasis * lightSampleRec.direction ), surfaceRec, sampleColor );
-								if ( lightMaterialPdf > 0.0 ) {
+								if ( lightMaterialPdf > 0.0 && ! any( isnan( sampleColor ) ) ) {
```
I'll see if I can get rid of the warnings by moving things around a bit. Observing the difference with and without the check make it abundantly clear that there are in fact NaN values at times. The warning itself is also quite funnily worded IMO. It _cannot_ be NaN, yet `isnan()` _may_ not be necessary.

mrxz replied:
Okay, I have the feeling ANGLE is doing some aggressive inlining. Placing the `isnan` check all the way into `specularColor` prevents the warnings, but makes communicating the result back quite ugly. Instead I've decided to not check for NaN values, but instead check the validity of the `sampleColor` by checking that all components (rgb) are >= 0.0, since we know negative color shouldn't occur.
```glsl
bool isValidSampleColor = all( greaterThanEqual( sampleColor, vec3( 0.0 ) ) );
```

gkjohnson replied:
It is a fairly odd warning... but thanks for taking a look at it! I was just as happy chalking it up to an ANGLE quirk!

bhouston commented:
Great work!  The efficiency gains are huge!

gkjohnson commented:
Phenomenal work @mrxz! Thanks for adding this!
